src/iceberg/literal.h

```diff
@@ -0,0 +1,138 @@
+  /// target type, it might be above the maximum allowed value for that type.
+  struct AboveMax {
+    bool operator==(const AboveMax&) const = default;
+    std::strong_ordering operator<=>(const AboveMax&) const = default;
+  };
+
+  using PrimitiveLiteralValue =
+      std::variant<bool,                     // for boolean
+                   int32_t,                  // for int, date
+                   int64_t,                  // for long, timestamp, timestamp_tz, time
+                   float,                    // for float
+                   double,                   // for double
+                   std::string,              // for string
+                   std::vector<uint8_t>,     // for binary, fixed
+                   std::array<uint8_t, 16>,  // for uuid and decimal
+                   BelowMin, AboveMax>;
+
+ public:
+  /// Factory methods for primitive types
+  static PrimitiveLiteral Boolean(bool value);
+  static PrimitiveLiteral Int(int32_t value);
+  static PrimitiveLiteral Long(int64_t value);
+  static PrimitiveLiteral Float(float value);
+  static PrimitiveLiteral Double(double value);
+  static PrimitiveLiteral String(std::string value);
+  static PrimitiveLiteral Binary(std::vector<uint8_t> value);
+
+  /// Create iceberg literal from bytes.
+  ///
+  /// See [this spec](https://iceberg.apache.org/spec/#binary-single-value-serialization)
+  /// for reference.
+  static Result<PrimitiveLiteral> Deserialize(std::span<const uint8_t> data);
+
+  /// Serialize iceberg literal to bytes.
+  ///
+  /// See [this spec](https://iceberg.apache.org/spec/#binary-single-value-serialization)
+  /// for reference.
+  Result<std::vector<uint8_t>> Serialize() const;
+
+  /// Get the Iceberg Type of the literal.
+  const std::shared_ptr<PrimitiveType>& type() const;
+
+  /// Converts this literal to a literal of the given type.
+  ///
+  /// When a predicate is bound to a concrete data column, literals are converted to match
+  /// the bound column's type. This conversion process is more narrow than a cast and is
+  /// only intended for cases where substituting one type is a common mistake (e.g. 34
+  /// instead of 34L) or where this API avoids requiring a concrete class (e.g., dates).
+  ///
+  /// If conversion to a target type is not supported, this method returns an error.
+  ///
+  /// This method may return BelowMin or AboveMax when the target type is not as wide as
+  /// the original type. These values indicate that the containing predicate can be
+  /// simplified. For example, std::numeric_limits<int>::max()+1 converted to an int will
+  /// result in AboveMax and can simplify a < std::numeric_limits<int>::max()+1 to always
+  /// true.
+  ///
+  /// \param target_type A primitive PrimitiveType
+  /// \return A Result containing a literal of the given type or an error if conversion
+  /// was not valid
+  Result<PrimitiveLiteral> CastTo(
+      const std::shared_ptr<PrimitiveType>& target_type) const;
```

Comment:
```suggestion
  Result<PrimitiveLiteral> CastTo(
      const PrimitiveType& target_type) const;
```

std::shared_ptr has a narrower use case

Reply:
We don't have a `Clone` for `PrimitiveType`?

Reply:
No, actually I believe most primitive types should use singleton.